cmd/manager/main.go

```diff
@@ -277,6 +278,23 @@ func main() {
 		setupLog.Error(err, "unable to create controller", "controller", "ClusterExtension")
 		os.Exit(1)
 	}
+
+	clusterCatalogFinalizers := crfinalizer.NewFinalizers()
+	if err := clusterCatalogFinalizers.Register(controllers.ClusterCatalogCacheDeletionFinalizer, finalizerFunc(func(ctx context.Context, obj client.Object) (crfinalizer.Result, error) {
+		return crfinalizer.Result{}, cacheFetcher.Remove(obj.GetName())
+	})); err != nil {
+		setupLog.Error(err, "unable to register finalizer", "finalizerKey", controllers.ClusterCatalogCacheDeletionFinalizer)
+		os.Exit(1)
+	}
+
+	if err = (&controllers.ClusterCatalogReconciler{
+		Client:     cl,
+		Finalizers: clusterCatalogFinalizers,
```

Comment:
I worry a little bit about using finalizers here. We'll end up with two different controllers that always race when catalogs are created and destroyed, which I'm guessing will add unnecessary reconcile churn.

WDYT about a best-effort garbage collector that runs the cleanup logic along the lines of:
```
if err := r.Get(req); errors.IsNotFound(err) {
    if err := cache.Cleanup(req); err != nil {
          metrics.IncrementCatalogCacheCleanupFailure()
          eventrecorder.Event("failed to clear catalog cache for %q", req)
    }
}
```

Another option is a garbage collector somewhat like catalogd has here: https://github.com/operator-framework/catalogd/blob/main/internal/garbagecollection/garbage_collector.go

Reply:
Can we still go with a "controller" approach that just doesn't modify the `ClusterCatalog` resource?

If the pod crashes, is our cache still populated? If so, we could do an initial list + cleanup on startup and then rely on informers to react to events that trigger our cleanup logic without modification of `ClusterCatalog` resources. Once the informer is up I am pretty confident it won't miss any events. Any events we miss are likely because the operator-controller manager container crashed and is going to restart and do the initial list+cleanup.

Does this sound like it could be a good middle ground?

Reply:
@everettraven I think this is the same as the best-effort option Joe suggested.

Anyway - I updated the controller to not change `ClusterCatalog`. Please take another look.


> If the pod crashes, is our cache still populated? If so, we could do an initial list + cleanup on startup and then rely on informers to react to events that trigger our cleanup logic without modification of `ClusterCatalog` resources. Once the informer is up I am pretty confident it won't miss any events. Any events we miss are likely because the operator-controller manager container crashed and is going to restart and do the initial list+cleanup.

On pod crash we will end up with filesystem containing cache since we are currently using `emptyDir` which persists between crashes. But [we will be rewriting](https://github.com/operator-framework/operator-controller/blob/df0e8481f09e6beac92d05c9ff62e31675d383b6/internal/catalogmetadata/cache/cache.go#L156-L158) the cache for each catalog since [this map](https://github.com/operator-framework/operator-controller/blob/df0e8481f09e6beac92d05c9ff62e31675d383b6/internal/catalogmetadata/cache/cache.go#L54) is not populated on restart. So it looks like using `emptyDir` is pointless here.

But my point was not about crashes: I was talking about issues when the manager is still running, but can't connect to API server. In this case it will still think that the cache is still present and valid because it hasn't seen delete event.

Perhaps getting rid of `emptyDir` will help?